Add AI skill for quarkus-hibernate-reactive (#54064)
quarkusio/quarkus: extensions/hibernate-reactive/deployment/src/main/resources/META-INF/quarkus-skill.md

    return sessionFactory.withTransaction(session ->
        session.find(Task.class, id)
            .onItem().ifNotNull().invoke(existing -> {
                existing.title = updated.title;
                existing.completed = updated.completed;
            })
    );
}

// DELETE
public Uni<Boolean> delete(Long id) {
    return sessionFactory.withTransaction(session ->
        session.find(Task.class, id)
            .onItem().ifNotNull().transformToUni(task ->
                session.remove(task).replaceWith(true))
            .onItem().ifNull().continueWith(false)
    );
}
```

### Key Patterns

- `withTransaction(session -> ...)` — automatically commits on success, rolls back on exception. Changes to managed entities are flushed automatically on commit.
- `session.persist(entity)` returns `Uni<Void>` — chain `.replaceWith(entity)` to get the entity back.
- `session.find(Type.class, id)` returns `Uni<T>` — may resolve to `null` if not found.

### Lazy Loading

Lazy initialization is not supported in Hibernate Reactive. Fetch lazy associations explicitly using `Mutiny.fetch()`:

```java
return sessionFactory.withTransaction(session ->
    session.find(Post.class, id)

>         .call(post -> Mutiny.fetch(post.getComments()))

Reviewer: Maybe this is not important, but I think it;s helpful to add the method signature here.

Because I don't think it's clear that this method will return a `Uni<Post>` with the collection initialized and not the initialized list of comments.

For reference, to return just the list of comments, it's possible to use `.chain` instead of `.call`:
```java
// Returns Uni<List<Comment>>
return sessionFactory.withTransaction(session ->
    session.find(Post.class, id)
        .chain(post -> Mutiny.fetch(post.getComments()))
```

Author: Good point — added comments explaining what `.call()` and `.chain()` return.

### Setup

Add `quarkus-hibernate-reactive` + a reactive SQL client (`reactive-pg-client`, `reactive-mysql-client`, etc.). Dev Services auto-starts a database container — no manual config needed in dev/test.

```properties
quarkus.hibernate-orm.schema-management.strategy=drop-and-create

> ```

Reviewer 2: This is not necessary: it's the default when using dev services.

```suggestion
```

It's also very, very bad advice in prod.

Author: Done — removed the db-kind config. Dev Services handles this automatically.

### Entities

Standard JPA annotations — same as Hibernate ORM:

```java
@Entity
public class Task {
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    public Long id;
    public String title;
    public boolean completed;
    public LocalDateTime createdAt;

    @PrePersist

>     void prePersist() { createdAt = LocalDateTime.now(); }

Reviewer 2: ```suggestion
    @CreationTimestamp
    public LocalDateTime createdAt;
```

Reviewer 3: Perhaps it's valuable to show how to declare hooks? But then yeah, it picked the wrong kind of example :-/

Author: Applied — using `@CreationTimestamp` now.

Author: Kept `@CreationTimestamp` since it's a more common pattern in entity definitions. Lifecycle callbacks could be a separate topic.

}
```

### SessionFactory API

Inject `Mutiny.SessionFactory` for all database operations:

```java
@Inject Mutiny.SessionFactory sessionFactory;
```

> Use `withSession` for reads, `withTransaction` for writes:

Reviewer 2: No, use `withTransaction` everywhere. `withSession` without a transaction only makes sense if you send a single SQL query.

And by the way... do we want to start recommending `@Transactional` @FroMage @lucamolteni ?

Reviewer 3: I was going to argue against `SessionFactory` in light of https://quarkus.io/guides/hibernate-reactive#using-transactional-with-hibernate-reactive mentionning we can just work on `Session`, indeed.

Hibernate ORM docs go directly for the `EntityManager`: https://quarkus.io/guides/hibernate-orm

NOTE: I thought we had pivoted to `Session` instead of `EntityManager` in Quarkus a while ago? I also don't see `StatelessSession` mentioned in the Quarkus ORM docs, that can't be right?

Reviewer 4: I'd start recommending it, yes

Reviewer 2: > NOTE: I thought we had pivoted to `Session` instead of `EntityManager` in Quarkus a while ago? I also don't see `StatelessSession` mentioned in the Quarkus ORM docs, that can't be right?

You're completely right. Our docs are very incomplete and in some cases out of date.

Author: Done — using `withTransaction` everywhere. Also added `@ReactiveTransactional` + `Session` as the primary/simpler approach, with `SessionFactory.withTransaction` as an alternative for explicit control.

Author: Good call — added `@ReactiveTransactional` + `Session` as the primary approach, with `SessionFactory.withTransaction` shown as an alternative.

Reviewer 2: `@ReactiveTransactional` has been deprecated for... years I think. There's been other annotations in the meantime, but nowadays you should use `@Transactional`. See https://quarkus.io/guides/hibernate-reactive#using-transactional-with-hibernate-reactive

```java
// CREATE
public Uni<Task> create(Task task) {
    return sessionFactory.withTransaction(session ->
        session.persist(task).replaceWith(task)
    );
}

// READ (single)
public Uni<Task> findById(Long id) {
    return sessionFactory.withSession(session ->
        session.find(Task.class, id)
    );
}

// READ (list)
public Uni<List<Task>> listAll() {
    return sessionFactory.withSession(session ->
        session.createQuery("FROM Task", Task.class).getResultList()
    );
}

// UPDATE
public Uni<Task> update(Long id, Task updated) {
    return sessionFactory.withTransaction(session ->
        session.find(Task.class, id)
            .onItem().ifNotNull().transformToUni(existing -> {
                existing.title = updated.title;
                existing.completed = updated.completed;

>                 return session.flush().replaceWith(existing);

Reviewer: No need to flush here because you are inside a transaction.
And you don't need to return a Uni if you just want to update the value:
```java
return sessionFactory.withTransaction( session -> session
					.find( GuineaPig.class, id )
					.onItem().ifNotNull().invoke( existing -> {
							existing.title = updated.title;
							existing.completed = updated.completed;
					} )
				);
```

Author: Done — removed the flush and simplified the update example per your suggestion.

            })
    );
}

// DELETE
public Uni<Boolean> delete(Long id) {
    return sessionFactory.withTransaction(session ->
        session.find(Task.class, id)
            .onItem().ifNotNull().transformToUni(task ->
                session.remove(task).replaceWith(true))
            .onItem().ifNull().continueWith(false)
    );
}
```

### Key Patterns

- `withTransaction(session -> ...)` — auto-commits on success, auto-rolls back on exception. Do NOT also use `@Transactional`.
- `withSession(session -> ...)` — read-only, no transaction management.
- `session.persist(entity)` returns `Uni<Void>` — chain `.replaceWith(entity)` to get the entity back.
- `session.flush()` is needed after modifying managed entities in a transaction. It's NOT automatic.
- `session.find(Type.class, id)` returns `Uni<T>` — may resolve to `null` if not found.

### Lazy Loading

Use `Mutiny.fetch()` to load lazy associations in reactive context:

```java
return sessionFactory.withSession(session ->
    session.find(Post.class, id)

>         .chain(post -> Mutiny.fetch(post.getComments()).replaceWith(post))

Reviewer: In this situation, one can use `.call`:
```java
session.find(...).call(post -> Mutiny.fetch(post.getComments()))
```

Author: Applied — showing both `.call()` (returns parent with collection initialized) and `.chain()` (returns the collection directly).

);
```

> Do NOT access lazy collections outside a session — it will fail silently or throw.

Reviewer 2: I would hope it always throws, if the collection is uninitialized? @DavideD can you confirm?

Reviewer: Yep, it will throw a `LazyInitializationException`. And in Hibernate Reactive lazy initialization is not supported anyway. One needs to fetch them explicitly: https://docs.hibernate.org/stable/reactive/reference/html_single/#_fetching_lazy_associations

Reviewer 3: What FUD 🙄

Author: Confirmed — the skill now states lazy init is not supported in HR and accessing unfetched collections always throws `LazyInitializationException`.

Author: Thanks — updated the skill to state this clearly and show `Mutiny.fetch()` with both `.call()` and `.chain()` patterns.

Author: Fixed — the skill now correctly states it always throws `LazyInitializationException` and lazy init is not supported in HR.

### Named Queries

```java
@Entity
@NamedQuery(name = "Task.findCompleted", query = "FROM Task WHERE completed = true")
public class Task { ... }

// Usage
session.createNamedQuery("Task.findCompleted", Task.class).getResultList()
```

### Testing

- Dev Services provides a real database — no mocking needed.
- Use `@QuarkusTest` normally with REST Assured.

> - Schema is recreated per test run with `drop-and-create`.

Reviewer 2: ```suggestion
- Schema is recreated per test run when using dev services.
```

Author: Applied.

> - For test isolation, either use `@TestTransaction` or manually clean data.

Reviewer 3: What's that? I can't find it from the ORM or HR docs 🤔

Author: Removed — that content is no longer in the skill.

### Common Pitfalls

> - Do NOT use `@Transactional` with `sessionFactory.withTransaction()` — they conflict.

Reviewer 3: Well, yeah, but then it also didn't use `@Transactional` :(

Author: Added `@ReactiveTransactional` as the primary pattern now.

> - `session.persist()` returns `Uni<Void>`, not the entity — use `.replaceWith(entity)`.

Reviewer 3: This is indeed annoying, but I've never found a satisfying way to provide this OOTB :(

Author: Acknowledged.

> - Always call `session.flush()` after modifying entity fields in `withTransaction` — changes are NOT auto-flushed.

Reviewer 3: They are not? I don't think that's true. If you commit the TX, changes are probably flushed, I can't imagine why not.

Reviewer: It's not true. No need to flush if they are using `.withTransaction`. I suggested a change to the example in another comment.

Author: Fixed — the skill now says "Changes to managed entities are flushed automatically on commit" and the examples don't include manual flush.

Author: Done — removed the flush, using the simplified update pattern you suggested.

- `session.find()` returns null (not an exception) for missing entities — handle with `.onItem().ifNull()`.

> - Lazy `@OneToMany` collections cause `LazyInitializationException` during JSON serialization — use `@JsonIgnore` on the collection field or eagerly fetch with `Mutiny.fetch()` before returning.

Reviewer 2: Or, preferably, use DTOs.

See also https://github.com/quarkusio/quarkus/issues/53263

cc @FroMage

Reviewer 3: Oh, I had completely forgotten I had a discussion with @gavinking about this and we sorta set a strategy for this. Let me see if I can find an issue I opened about it…

Reviewer 3: Oh, you linked to it 🤣

Author: Added DTO mention to the pitfalls section.

Author: The DTO recommendation is now in the pitfalls section, with a reference to issue #53263.

Author: 😄

        });
}

@Transactional
public Uni<Boolean> delete(Long id) {
    return session.find(Task.class, id)
        .onItem().ifNotNull().transformToUni(task ->
            session.remove(task).replaceWith(true))
        .onItem().ifNull().continueWith(false);
}
```

### SessionFactory API (alternative)

Use `Mutiny.SessionFactory` when you need explicit transaction control:

```java
@Inject Mutiny.SessionFactory sessionFactory;

public Uni<Task> create(Task task) {
    return sessionFactory.withTransaction(session ->
        session.persist(task).replaceWith(task)
    );
}

public Uni<Task> update(Long id, Task updated) {
    return sessionFactory.withTransaction(session ->
        session.find(Task.class, id)
            .onItem().ifNotNull().invoke(existing -> {
                existing.title = updated.title;
                existing.completed = updated.completed;
            })
    );
}
```

### Key Patterns

- `withTransaction(session -> ...)` — automatically commits on success, rolls back on exception. Changes to managed entities are flushed automatically on commit.
- `session.persist(entity)` returns `Uni<Void>` — chain `.replaceWith(entity)` to get the entity back.

> - `session.find(Type.class, id)` returns `Uni<T>` — may resolve to `null` if not found.

Reviewer 2: ```suggestion

- `@Inject Mutiny.Session` + `@Transactional` (preferably) or `@Inject Mutiny.SessionFactory` + `sessionFactory.withTransaction(session -> ...)` (legacy code or advanced usage) — automatically commits on success, rolls back on exception. Changes to managed entities are flushed automatically on commit.
- `session.persist(entity)` returns `Uni<Void>` — chain `.replaceWith(entity)` to get the entity back.
- `session.find(Type.class, id)` returns `Uni<Type>` — may resolve to `null` if not found.
```

### Lazy Loading

Lazy initialization is not supported in Hibernate Reactive. Fetch lazy associations explicitly using `Mutiny.fetch()`:

```java
// Returns Uni<Post> with comments initialized
return sessionFactory.withTransaction(session ->
    session.find(Post.class, id)
        .call(post -> Mutiny.fetch(post.getComments()))

> );

Reviewer 2: ```suggestion
return session.find(Post.class, id)
    .call(post -> Mutiny.fetch(post.getComments()));
```

// Returns Uni<List<Comment>> directly

> return sessionFactory.withTransaction(session ->

Reviewer 3: Same here

    session.find(Post.class, id)
        .chain(post -> Mutiny.fetch(post.getComments()))
);
```

- `.call()` — fetches the association, returns the parent entity (Post) with the collection initialized.
- `.chain()` — fetches the association, returns the fetched collection (List\<Comment>) directly.

Accessing an unfetched lazy collection throws `LazyInitializationException`.

### Named Queries

```java
@Entity
@NamedQuery(name = "Task.findCompleted", query = "FROM Task WHERE completed = true")
public class Task { ... }

// Usage
session.createNamedQuery("Task.findCompleted", Task.class).getResultList()
```

### Testing

> - Dev Services provides a real database — no mocking needed.

Reviewer 2: ```suggestion
- Dev Services provide a real database — no mocking needed.
```

- Use `@QuarkusTest` normally with REST Assured.

> - Schema is recreated per test run when using Dev Services.

Reviewer 2: ```suggestion
- Schema and test data are recreated per test run (but shared between test classes) when using Dev Services.
```
